runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -508,6 +530,12 @@ private void checkTimestamp(Instant timestamp) {
     /** The pane of the current element. */
     private final PaneInfo pane;
 
+    /**
+     * The state namespace for this context. Lazily initialized; will crash if access when more than
```

Comment:
`// This should never happen`

More specifically, A `DoFn` that accesses a `StateNamespace` should always have `observesWindow` as `true`, and as a result should always explode windows before passing elements to the `@ProcessElement` method.